src/io/iohk/decco/PartialCodec.scala

```diff
@@ -0,0 +1,82 @@
+package io.iohk.decco
+
+import java.nio.ByteBuffer
+
+import scala.reflect.runtime.universe.TypeTag
+
+trait PartialCodec[T] { self =>
+
+  import PartialCodec._
+
+  def size(t: T): Int
+  def typeCode: String
+  def encode(t: T, start: Int, destination: ByteBuffer): Unit
+  def decode(start: Int, source: ByteBuffer): Either[Failure, DecodeResult[T]]
+
+  // SOME COMPOSITION FUNCTIONS
+
+  final def zip[U](that: PartialCodec[U]): PartialCodec[(T, U)] = new PartialCodec[(T, U)] {
+
+    def size(tu: (T, U)): Int = tu match {
+      case (t, u) => self.size(t) + that.size(u)
+    }
+
+    def decode(start: Int, source: ByteBuffer): Either[Failure, DecodeResult[(T, U)]] = {
+      self.decode(start, source) match {
```

Comment:
Looks like this could be rewritten to remove redundancy and nesting:
```
  self
    .decode(start, source)
    .flatMap { case DecodeResult(t, nextIndex) =>
      (t, that.decode(nextIndex, source))
    }
    .map { case (t, DecodeResult(u, resultingNextIndex)) =>
      DecodeResult((t, u), resultingNextIndex)
    }
```